siliconcompiler/core.py

```diff
@@ -4090,210 +4053,286 @@ def run(self):
+        cur_steplist = self.get('option', 'steplist')
+        pre_remote_steplist = {
+            'steplist': cur_steplist,
+            'set': self.schema._is_set(self.schema._search('option', 'steplist')),
+        }
+        remote_preprocess(self, steplist)
 
-                    if (self._get_in_job(step, index) != jobname):
-                        # If we specify a different job as input to this task,
-                        # we assume we are good to run it.
-                        tasks_to_run[nodename] = []
-                    else:
-                        tasks_to_run[nodename] = inputs
-
-                    processes[nodename] = multiprocessor.Process(target=self._runtask,
-                                                                 args=(step, index, status))
-
-            running_tasks = []
-            while len(tasks_to_run) > 0 or len(running_tasks) > 0:
-                # Check for new tasks that can be launched.
-                for node, deps in list(tasks_to_run.items()):
-                    # TODO: breakpoint logic:
-                    # if task is breakpoint, then don't launch while len(running_tasks) > 0
-
-                    dep_was_successful = False
-                    had_deps = len(deps) > 0
-                    tool, task = self._get_tool_task(
-                        processes[node]._args[0],
-                        processes[node]._args[1])
-
-                    # Clear any tasks that have finished from dependency list.
-                    for in_node in deps.copy():
-                        if status[in_node] != TaskStatus.PENDING:
-                            deps.remove(in_node)
-                        if status[in_node] == TaskStatus.SUCCESS:
-                            dep_was_successful = True
-                        if status[in_node] == TaskStatus.ERROR:
-                            # Fail if any dependency failed for non-builtin task
-                            if not self._is_builtin(tool, task):
-                                status[node] = TaskStatus.ERROR
-                                break
-
-                    # Fail if no dependency successfully finished for builtin task
-                    if had_deps and len(deps) == 0 \
-                            and self._is_builtin(tool, task) and not dep_was_successful:
+        # Run the job on the remote server, and wait for it to finish.
+        # Set logger to indicate remote run
+        self._init_logger(step='remote', index='0', in_run=True)
+        remote_run(self)
+
+        # Delete the job's data from the server.
+        delete_job(self)
+        # Restore logger
+        self._init_logger(in_run=True)
+        # Restore steplist
+        if pre_remote_steplist['set']:
+            self.set('option', 'steplist', pre_remote_steplist['steplist'])
+
+    def _prepare_tasks(self, tasks_to_run, processes, flow, status, steplist, indexlist):
```

Comment:
I think this next set of three should be renamed: prepare_task, check_node_dependencies, and lanch_task makes it hard to follow the naming, you are preparing a node, so lets keep that naming pattern at least.

Reply:
I think that was part of another refactor Noah suggested as well.
Did that now.
Should I also rename `TaskStatus` to `Node Status`?

Reply:
NodeStatus is fine